README.md
# arcadeflex056
Port of mame 0.56 to java language

Direct port of mame 0.56 from c to java language


changelog.txt
17/05/2019
==========
-Added rollrace working driver
-Added wiz working driver
-Added kangaroo working driver
-Added strnskil working driver
-Added timelimt working driver
-Added thepit working driver

12/05/2019
==========
-Added working naughtyb and vigilant working drivers
-Added segar WIP driver (NO SOUND)

10/05/2019
==========
-Added troangel driver (fully playable)
-Added yard driver (fully playable)
-Added WIP mcr3 driver (non working driver still)

09/05/2019
==========
-Correct colors in snk, taitosj and hal21 drivers

08/05/2019
==========
-Added support for triviaes game (from mame 0.104u7) in balsente driver
-All games of m62 driver are fully playable
-Added mpatrol driver
-Added kungfum game in m62 driver (WIP)

06/05/2019
==========
-video border issue solved (e.g. in swimmer driver)
-kingobox is fully playable (moved to mame056 branch)
-Added kingobox driver
-coin_counter_w method is now a WriteHandlerPtr (many classes have been updated)

03/05/2019
=========
-Added sauro fully playable driver
-Added WIP speedbal driver

30/04/2019
==========
-Added arabian driver (fully playable)

28/04/2019
==========
-Added exidy440 working driver (sound WIP)

25/04/2019
==========
-Moved WIP drivers to WIP folder
-Added WIP polepos driver

24/04/2019
==========
-Offscreen drawing bug fixed
-Added locomotn driver (fully playable all games)
-Added mappy driver (fully playable all games)
-Added grobda driver (fully playable all games)
-Added cosmic driver (fully playable games)
-Added WIP commando and balsente drivers
-Implemented strstr in cstring
-Implemented cpu_getscanlineperiod in cpuexec

22/04/2019
==========
-Solved palette colors problem!!! (now drivers like system1, jackal, etc. work now)
-Added retofinv driver (retofin1 and retofin2 drivers fully playable)

21/04/2019
==========
-Added ironhors driver (fully playable)
-Added tutankhm driver (fully playable)
-Added junofrst driver (some gfx issues)

20/04/2019
==========
-Added i8039 sound to vicdual driver

19/04/2019
==========
-Added install_port_write_handler method in memory class

18/04/2019
==========
-brkthru game works perfectly now!
-Added ROM_LOAD16_BYTE functionality in common class (now jackal shows the sprites, not the background)
-Added mario driver (no sound/fully playable/correct colors)
-Added superpac driver (fully playable)
-Added sidearms driver (imperfect colors/no sound)
-Video problem with the first pixel column solved

16/04/2019
==========
-Added marineb driver (fully playable)
-Added espial driver (fully playable)
-Added champbas driver (fully playable)
-Added jackal driver (WIP)
-Added sgladiat driver (fully playable)
-Added jack driver (fully playable)
-Added shisen driver (NO SOUND still)
-Added jailbrek driver (full playable)
-Added yiear driver (full playable)
-Added support for ym3812 and ym3526
-Added support for ym8950

15/04/2019
==========
-Added support for ym2203 sound driver (0.56) using fm classes in 0.37b11
-Added sound to Gunsmoke driver

09/04/2019
==========
-irobot driver (WIP/many thing to be done)
-Added hal21 driver (WIP)

08/04/2019
==========
-Added docastle driver (WIP)

07/04/2019
==========
-Added MSM5205 sound and added sound to Kchamp driver
-Added preliminary version of snk driver (WIP)
-Added preliminary version of hal21 driver (WIP)
-Added ladybug driver (Full Playable)
-Solved speed video issues

06/04/2019
==========
-Added preliminary version of tnzs driver (WIP)
-Added tp84 driver (fully playable)

05/04/2019
==========
-Video class 0.56 fully ported!

04/04/2019
==========
-More functions implemented in video 056 class

03/04/2019
==========
-Added hyperspt driver fully playable (partial sound)
-Solved all video problems in taitosj driver
-Added popeye and tehkanwc (WIP/in progess)
-Solved all video problems in fastfred driver

02/04/2019
==========
-Added bosco driver (fully playable)
-Added gunsmoke driver (No Sound/fully playable)
-Added trackfld driver

01/04/2019
==========
-Starting porting of video 056

27/03/2019
==========
-Added WIP taitosj driver

26/03/2019
==========
-Added pbaction driver
-Some adjustments in kchamp, fastfred ans system1 drivers
-Added sonson driver

25/03/2019
==========
-Added rallyx driver
-Added vicdual driver
-Fixed bug when sample file doesn't exists

24/03/2019
==========
-Ported File functions to cstdio common
-Fixed kingball in galaxian driver
-Fixed mooncrgx in galaxian driver (roms is not encrypted ,was wrong in mame as well)
-Added working gaplus driver

22/03/2019
==========
-Added sega vector games (WIP/NOT WORKING still)

21/03/2019
==========
-First version of starwars and gberet (WIP/still NOT WORKING) for testing vector graphics and tiles
-First implementation of tilemapC (WIP/NOT WORKING still)

19/03/2019
==========
-Ported berzerk driver (SOUND not ported still)
-Berzerk now has sound and it's fully playable

16/03/2019
==========
Fixed galaga driver

15/03/2019
==========
-Ported to 0.56 vector components and asteroid driver (WIP)
-Ported to 0.56 vector and vectorH classes
-Added shaolins driver (FULLY PLAYABLE)

14/03/2019
==========
-Added MIKIE driver (fully playable)
-Added WIP galaga and fastfred drivers
-Added rocnrope driver (fully playable) and kchamp driver (doesn't shows the menu; same ink for back and foreground)
-Added cclimber driver (games supported: cclimber, cclimbrj, ccboot, ccboot2, ckong, ckonga, ckongjeu, ckongo, 
ckongalc, monkeyd, rpatrolb, silvland, swimmer, swimmera, swimmerb, guzzler)
-M6809 cpu added (but we don't have a game that uses it yet)

13/03/2019
==========
-Fixed samples playing (using old mixer from 0.37b11. Still have to figure out why the new mixer isn't working properly)
-Added sprite case for newsin7 in galaxian vidhrdw
-Added bombjack driver to fully playable 
-Added frogger driver to fully playable
-galaxian driver compatibility updated
-Workaround for double window intialaze in video.java (fixed some galaxian games)

emulator/src/mame056/inputH.java
/**
 * Ported to 0.56
 */
package mame056;

import static mame056.input.*;

public class inputH {

    public static class KeyboardInfo {

        public KeyboardInfo(String name, int code, int standardcode) {
            this.name = name;
            this.code = code;
            this.standardcode = standardcode;
        }
        public String name;/* OS dependant name; 0 terminates the list */
        public int code;/* OS dependant code */
        public int standardcode;/* CODE_xxx equivalent from list below, or CODE_OTHER if n/a */
    }
    /*TODO*///struct JoystickInfo
/*TODO*///{
/*TODO*///	char *name; /* OS dependant name; 0 terminates the list */
/*TODO*///	unsigned code; /* OS dependant code */
/*TODO*///	InputCode standardcode;	/* CODE_xxx equivalent from list below, or CODE_OTHER if n/a */
/*TODO*///};
/*TODO*///
    /* key */
    public static final int KEYCODE_A = 0;
    public static final int KEYCODE_B = 1;
    public static final int KEYCODE_C = 2;
    public static final int KEYCODE_D = 3;
    public static final int KEYCODE_E = 4;
    public static final int KEYCODE_F = 5;
    public static final int KEYCODE_G = 6;
    public static final int KEYCODE_H = 7;
    public static final int KEYCODE_I = 8;
    public static final int KEYCODE_J = 9;
    public static final int KEYCODE_K = 10;
    public static final int KEYCODE_L = 11;
    public static final int KEYCODE_M = 12;
    public static final int KEYCODE_N = 13;
    public static final int KEYCODE_O = 14;
    public static final int KEYCODE_P = 15;
    public static final int KEYCODE_Q = 16;
    public static final int KEYCODE_R = 17;
    public static final int KEYCODE_S = 18;
    public static final int KEYCODE_T = 19;
    public static final int KEYCODE_U = 20;
    public static final int KEYCODE_V = 21;
    public static final int KEYCODE_W = 22;
    public static final int KEYCODE_X = 23;
    public static final int KEYCODE_Y = 24;
    public static final int KEYCODE_Z = 25;
    public static final int KEYCODE_0 = 26;
    public static final int KEYCODE_1 = 27;
    public static final int KEYCODE_2 = 28;
    public static final int KEYCODE_3 = 29;
    public static final int KEYCODE_4 = 30;
    public static final int KEYCODE_5 = 31;
    public static final int KEYCODE_6 = 32;
    public static final int KEYCODE_7 = 33;
    public static final int KEYCODE_8 = 34;
    public static final int KEYCODE_9 = 35;
    public static final int KEYCODE_0_PAD = 36;
    public static final int KEYCODE_1_PAD = 37;
    public static final int KEYCODE_2_PAD = 38;
    public static final int KEYCODE_3_PAD = 39;
    public static final int KEYCODE_4_PAD = 40;
    public static final int KEYCODE_5_PAD = 41;
    public static final int KEYCODE_6_PAD = 42;
    public static final int KEYCODE_7_PAD = 43;
    public static final int KEYCODE_8_PAD = 44;
    public static final int KEYCODE_9_PAD = 45;
    public static final int KEYCODE_F1 = 46;
    public static final int KEYCODE_F2 = 47;
    public static final int KEYCODE_F3 = 48;
    public static final int KEYCODE_F4 = 49;
    public static final int KEYCODE_F5 = 50;
    public static final int KEYCODE_F6 = 51;
    public static final int KEYCODE_F7 = 52;
    public static final int KEYCODE_F8 = 53;
    public static final int KEYCODE_F9 = 54;
    public static final int KEYCODE_F10 = 55;
    public static final int KEYCODE_F11 = 56;
    public static final int KEYCODE_F12 = 57;
    public static final int KEYCODE_ESC = 58;
    public static final int KEYCODE_TILDE = 59;
    public static final int KEYCODE_MINUS = 60;
    public static final int KEYCODE_EQUALS = 61;
    public static final int KEYCODE_BACKSPACE = 62;
    public static final int KEYCODE_TAB = 63;
    public static final int KEYCODE_OPENBRACE = 64;
    public static final int KEYCODE_CLOSEBRACE = 65;
    public static final int KEYCODE_ENTER = 66;
    public static final int KEYCODE_COLON = 67;
    public static final int KEYCODE_QUOTE = 68;
    public static final int KEYCODE_BACKSLASH = 69;
    public static final int KEYCODE_BACKSLASH2 = 70;
    public static final int KEYCODE_COMMA = 71;
    public static final int KEYCODE_STOP = 72;
    public static final int KEYCODE_SLASH = 73;
    public static final int KEYCODE_SPACE = 74;
    public static final int KEYCODE_INSERT = 75;
    public static final int KEYCODE_DEL = 76;
    public static final int KEYCODE_HOME = 77;
    public static final int KEYCODE_END = 78;
    public static final int KEYCODE_PGUP = 79;
    public static final int KEYCODE_PGDN = 80;
    public static final int KEYCODE_LEFT = 81;
    public static final int KEYCODE_RIGHT = 82;
    public static final int KEYCODE_UP = 83;
    public static final int KEYCODE_DOWN = 84;
    public static final int KEYCODE_SLASH_PAD = 85;
    public static final int KEYCODE_ASTERISK = 86;
    public static final int KEYCODE_MINUS_PAD = 87;
    public static final int KEYCODE_PLUS_PAD = 88;
    public static final int KEYCODE_DEL_PAD = 89;
    public static final int KEYCODE_ENTER_PAD = 90;
    public static final int KEYCODE_PRTSCR = 91;
    public static final int KEYCODE_PAUSE = 92;
    public static final int KEYCODE_LSHIFT = 93;
    public static final int KEYCODE_RSHIFT = 94;
    public static final int KEYCODE_LCONTROL = 95;
    public static final int KEYCODE_RCONTROL = 96;
    public static final int KEYCODE_LALT = 97;
    public static final int KEYCODE_RALT = 98;
    public static final int KEYCODE_SCRLOCK = 99;
    public static final int KEYCODE_NUMLOCK = 100;
    public static final int KEYCODE_CAPSLOCK = 101;
    public static final int KEYCODE_LWIN = 102;
    public static final int KEYCODE_RWIN = 103;
    public static final int KEYCODE_MENU = 104;

    public static final int __code_key_first = KEYCODE_A;
    public static final int __code_key_last = KEYCODE_MENU;

    /* joy */
    public static final int JOYCODE_1_LEFT = 105;
    public static final int JOYCODE_1_RIGHT = 106;
    public static final int JOYCODE_1_UP = 107;
    public static final int JOYCODE_1_DOWN = 108;
    public static final int JOYCODE_1_BUTTON1 = 109;
    public static final int JOYCODE_1_BUTTON2 = 110;
    public static final int JOYCODE_1_BUTTON3 = 111;
    public static final int JOYCODE_1_BUTTON4 = 112;
    public static final int JOYCODE_1_BUTTON5 = 113;
    public static final int JOYCODE_1_BUTTON6 = 114;
    public static final int JOYCODE_2_LEFT = 115;
    public static final int JOYCODE_2_RIGHT = 116;
    public static final int JOYCODE_2_UP = 117;
    public static final int JOYCODE_2_DOWN = 118;
    public static final int JOYCODE_2_BUTTON1 = 119;
    public static final int JOYCODE_2_BUTTON2 = 120;
    public static final int JOYCODE_2_BUTTON3 = 121;
    public static final int JOYCODE_2_BUTTON4 = 122;
    public static final int JOYCODE_2_BUTTON5 = 123;
    public static final int JOYCODE_2_BUTTON6 = 124;
    public static final int JOYCODE_3_LEFT = 125;
    public static final int JOYCODE_3_RIGHT = 126;
    public static final int JOYCODE_3_UP = 127;
    public static final int JOYCODE_3_DOWN = 128;
    public static final int JOYCODE_3_BUTTON1 = 129;
    public static final int JOYCODE_3_BUTTON2 = 130;
    public static final int JOYCODE_3_BUTTON3 = 131;
    public static final int JOYCODE_3_BUTTON4 = 132;
    public static final int JOYCODE_3_BUTTON5 = 133;
    public static final int JOYCODE_3_BUTTON6 = 134;
    public static final int JOYCODE_4_LEFT = 135;
    public static final int JOYCODE_4_RIGHT = 136;
    public static final int JOYCODE_4_UP = 137;
    public static final int JOYCODE_4_DOWN = 138;
    public static final int JOYCODE_4_BUTTON1 = 139;
    public static final int JOYCODE_4_BUTTON2 = 140;
    public static final int JOYCODE_4_BUTTON3 = 141;
    public static final int JOYCODE_4_BUTTON4 = 142;
    public static final int JOYCODE_4_BUTTON5 = 143;
    public static final int JOYCODE_4_BUTTON6 = 144;

    public static final int __code_joy_first = JOYCODE_1_LEFT;
    public static final int __code_joy_last = JOYCODE_4_BUTTON6;

    public static final int __code_max = 145;/* Temination of standard code */

 /* special */
    public static final int CODE_NONE = 0x8000;/* no code, also marker of sequence end */
    public static final int CODE_OTHER = 0x8001;/* OS code not mapped to any other code */
    public static final int CODE_DEFAULT = 0x8002;/* special for input port definitions */
    public static final int CODE_PREVIOUS = 0x8003;/* special for input port definitions */
    public static final int CODE_NOT = 0x8004;/* operators for sequences */
    public static final int CODE_OR = 0x8005;/* operators for sequences */


 /* Wrapper for compatibility */
 /*TODO*///#define KEYCODE_OTHER CODE_OTHER
/*TODO*///#define JOYCODE_OTHER CODE_OTHER
/*TODO*///#define KEYCODE_NONE CODE_NONE
/*TODO*///#define JOYCODE_NONE CODE_NONE

    /*TODO*////* Wrappers for compatibility */
/*TODO*///#define keyboard_name                   code_name
/*TODO*///#define keyboard_pressed                code_pressed
    public static int keyboard_pressed_memory(int code) {
        return code_pressed_memory(code);
    }
    /*TODO*///#define keyboard_pressed_memory_repeat  code_pressed_memory_repeat
/*TODO*///#define keyboard_read_async             code_read_async
    /**
     * Sequence code funtions
     */
    /* NOTE: If you modify this value you need also to modify the SEQ_DEF declarations */
    public static final int SEQ_MAX = 16;

    public static int seq_get_1(int[] a) {
        return a[0];
    }

    /* NOTE: It's very important that this sequence is EXACLY long SEQ_MAX */
    public static int[] SEQ_DEF_6(int a, int b, int c, int d, int e, int f) {
        return new int[]{a, b, c, d, e, f, CODE_NONE, CODE_NONE, CODE_NONE, CODE_NONE, CODE_NONE, CODE_NONE, CODE_NONE, CODE_NONE, CODE_NONE, CODE_NONE};
    }

    public static int[] SEQ_DEF_5(int a, int b, int c, int d, int e) {
        return SEQ_DEF_6(a, b, c, d, e, CODE_NONE);
    }

    public static int[] SEQ_DEF_4(int a, int b, int c, int d) {
        return SEQ_DEF_5(a, b, c, d, CODE_NONE);
    }

    public static int[] SEQ_DEF_3(int a, int b, int c) {
        return SEQ_DEF_4(a, b, c, CODE_NONE);
    }

    public static int[] SEQ_DEF_2(int a, int b) {
        return SEQ_DEF_3(a, b, CODE_NONE);
    }

    public static int[] SEQ_DEF_1(int a) {
        return SEQ_DEF_2(a, CODE_NONE);
    }

    public static int[] SEQ_DEF_0() {
        return SEQ_DEF_1(CODE_NONE);
    }
}
